python/lsst/base/packages.py

```diff
@@ -0,0 +1,258 @@
+   already been imported.
+3. Other packages provide no run-time accessible version information (e.g., astrometry_net): we get their
+   version from interrogating the environment. Currently, that means EUPS; if EUPS is replaced or dropped then
+   we'll need to consider an alternative means of getting this version information.
+4. Local versions of packages (a non-installed EUPS package, selected with "setup -r /path/to/package"): we
+   identify these through the environment (EUPS again) and use as a version the path supplemented with the
+   git SHA and, if the git repo isn't clean, an MD5 of the diff.
+
+These package versions are collected and stored in a Packages object, which provides useful comparison and
+persistence features.
+
+Example usage:
+
+    from lsst.base import Packages
+    pkgs = Packages.fromSystem()
+    print "Current packages:", pkgs
+    old = Packages.read("/path/to/packages.pickle")
+    print "Old packages:", old
+    print "Missing packages compared to before:", pkgs.missing(old)
+    print "Extra packages compared to before:", pkgs.extra(old)
+    print "Different packages: ", pkgs.difference(old)
+    old.update(pkgs)  # Include any new packages in the old
+    old.write("/path/to/packages.pickle")
+"""
+
+import os
+import sys
+import hashlib
+import importlib
+import subprocess
+import cPickle as pickle
+from collections import Mapping
+
+from .baseLib import getRuntimeVersions
+
+__all__ = ["getVersionFromPythonModule", "getPythonPackages", "getEnvironmentPackages", "Packages"]
+
+
+# Packages used at build-time (e.g., header-only)
+BUILDTIME = set(["boost", "eigen", "tmv"])
+
+# Python modules to attempt to load so we can try to get the version
+# We do this because the version only appears to be available from python, but we use the library
+PYTHON = set(["galsim"])
+
+# Packages that don't seem to have a mechanism for reporting the runtime version
+# We need to guess the version from the environment
+ENVIRONMENT = set(["astrometry_net", "astrometry_net_data", "minuit2", "xpa"])
+
+
+def getVersionFromPythonModule(module):
+    """Determine the version of a python module
+
+    Will raise AttributeError if the __version__ attribute is not set.
+
+    We supplement the version with information from the __dependency_versions__
+    (a specific variable set by LSST's sconsUtils at build time) only for packages
+    that are typically used only at build-time.
+    """
+    version = module.__version__
+    if hasattr(module, "__dependency_versions__"):
+        # Add build-time dependencies
+        deps = module.__dependency_versions__
+        buildtime = BUILDTIME & set(deps.iterkeys())
+        if buildtime:
+            version += " with " + " ".join("%s=%s" % (pkg, deps[pkg]) for pkg in buildtime)
+    return version
+
+def getPythonPackages():
+    """Provide a dict of imported python packages and their versions
```

Comment:
First time I read this I thought it was telling me that I was supposed to give a dict as an argument to the function. I think using "Provides" rather than "Provide" would avoid that confusion, or alternatively "Return" if the tense is a problem..